### PR TITLE
fix: check all examples for margin key in DataCollatorForPreference

### LINKED ISSUE #5539: fix: DataCollatorForRewardModelingDataset checks margin only on examples[0], causing silent data loss after shuffle
## Problem

In `trl/trainer/reward_trainer.py`, `DataCollatorForRewardModelingDataset.torch_call()` checks for the `"margin"` key only on `examples[0]`:

```python
if "margin" in examples[0]:
    margins = torch.tensor([example["margin"] for example in examples], dtype=torch.float)
```

This has two failure modes:

1. **Silent data loss**: After dataset shuffling, if the first example lacks `"margin"` but subsequent examples have it, margins are silently dropped and the margin-aware reward loss is skipped with no warning.
2. **Runtime `KeyError`**: If the first example has `"margin"` but a later one does not (inconsistent dataset), the list comprehension crashes.

## Expected Behavior

All examples should be checked for the `"margin"` key. Missing margins should default to `0.0` gracefully.

## Fix

```python
if any("margin" in example for example in examples):
    margins = torch.tensor([example.get("margin", 0.0) for example in examples], dtype=torch.float)
```

### PR DESCRIPTION
**Pull request by CrepuscularIRIS**
# What does this PR do?

Fixes the margin check in `DataCollatorForPreference.torch_call()` which only inspected `examples[0]` for the `"margin"` key. This caused two failure modes:

1. **Silent data loss**: After dataset shuffling, if the first example in a batch lacks `"margin"` but others have it, all margins are silently dropped — the margin-aware reward loss is skipped with no warning.
2. **Runtime KeyError**: If the first example has `"margin"` but a later one does not (inconsistent dataset), the list comprehension crashes.

**Fix**: Use `any()` to check all examples and `.get("margin", 0.0)` for graceful defaults.

Fixes #5539

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Low Risk**
> Small, localized change to collator batching logic with added tests; risk is limited to margin-handling behavior in reward training batches.
> 
> **Overview**
> Fixes `DataCollatorForPreference.torch_call()` to detect `margin` across *all* batch examples (not just `examples[0]`) and to default missing margins to `0.0`, preventing silent margin drops and `KeyError`s on inconsistent batches.
> 
> Adds two unit tests covering mixed batches where `margin` appears only in some examples (including shuffled order) to ensure `margin` is always returned with appropriate defaults.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9d2d7616a4a1ddabe58f4622f4c0f1c01a34cb1b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by CrepuscularIRIS** [comment]
Thanks for the review. Good catch on the DPO trainer — I'll update the PR to propagate the fix to `dpo_trainer.py` as well and deduplicate the `any()` check into a stored variable.

**Comment by CrepuscularIRIS** [comment]
Pushed the follow-up commit:

- Stored `has_margin` result to avoid the duplicate `any()` iteration
- Propagated the same `any()` + `.get()` fix to `dpo_trainer.py` for `ref_chosen_logps` and `ref_rejected_logps` (same `examples[0]` anti-pattern)

Both issues from the review should now be addressed.

**Comment by CrepuscularIRIS** [comment]
Closing — I missed that @antoinsader's #5562 was opened at 2026-04-15T20:48:36Z, about 13 minutes before mine. Yielding so they can finish unencumbered. Apologies for the noise.

One note for @antoinsader in case it's useful: while reviewing feedback on this PR I noticed the same `examples[0]`-only pattern also affects `trl/trainer/dpo_trainer.py` for `ref_chosen_logps` / `ref_rejected_logps` — not just `reward_trainer.py`. Feel free to fold that into #5562 or #5564 if the maintainer decides the fix is warranted.

Re @qgallouedec's question ("any example of a paper/implementation where some samples have a margin and some not?") — I don't have one; happy to defer to your judgment on whether the fix is even needed.

**Comment by antoinsader** [comment]
@CrepuscularIRIS  thank you for referring ref_chosen_logps / ref_rejected_logps, I will make sure to check them and fold them into my PR.
